src/Sylius/Component/Resource/Repository/InMemoryRepository.php

```diff
@@ -0,0 +1,245 @@
+        }
+
+        $this->interface = $interface;
+        $this->accessor = PropertyAccess::createPropertyAccessor();
+        $this->arrayObject = new ArrayObject();
+    }
+
+    /**
+     * @param ResourceInterface $resource
+     *
+     * @throws \InvalidArgumentException
+     * @throws UnexpectedTypeException
+     */
+    public function add(ResourceInterface $resource)
+    {
+        if (!in_array($this->interface, class_implements($resource))) {
+            throw new UnexpectedTypeException($resource, $this->interface);
+        }
+
+        if (null === $resource->getId()) {
+            throw new \InvalidArgumentException('Resource\'s id needs to be set in order to add.');
+        }
+
+        if ($this->arrayObject->offsetExists($resource->getId())) {
+            throw new \InvalidArgumentException(
+                sprintf('An object with id \'%s\' is already in the repository.', $resource->getId())
+            );
+        }
+
+        $this->arrayObject->offsetSet($resource->getId(), $resource);
+    }
+
+    /**
+     * @param ResourceInterface $resource
+     */
+    public function remove(ResourceInterface $resource)
+    {
+        if ($this->arrayObject->offsetExists($resource->getId())) {
+            $this->arrayObject->offsetUnset($resource->getId());
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function find($id)
+    {
+        if (empty($id)) {
+            return null;
+        }
+
+        if ($this->arrayObject->offsetExists($id)) {
+            return $this->arrayObject->offsetGet($id);
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findAll()
+    {
+        return array_values($this->arrayObject->getArrayCopy());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findBy(array $criteria = array(), array $orderBy = null, $limit = null, $offset = null)
+    {
+        $results = $this->arrayObject->getArrayCopy();
+
+        if (isset($criteria)) {
+            $results = $this->applyCriteria($results, $criteria);
+        }
+
+        if (isset($orderBy)) {
+            $results = $this->applyOrder($results, $orderBy);
+        }
+
+        $results = array_slice($results, $offset, $limit);
+
+        return $results;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \UnexpectedValueException
+     */
+    public function findOneBy(array $criteria)
+    {
+        $results = $this->applyCriteria($this->findAll(), $criteria);
+
+        if (1 < $length = count($results)) {
+            throw new \UnexpectedValueException(sprintf('Non unique result. Number of results: %s.', $length));
+        }
+
+        return $results[0];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getClassName()
+    {
+        return $this->interface;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createPaginator(array $criteria = null, array $orderBy = null)
```

Comment:
Not sure about `array $variable = null`, why not `array $variable = []`? There shouldn't be performance difference between `isset` and `empty` language constructs.